File: compat/suites/python-sdk/groups/lambda_.py
"""
groups/lambda_.py — Lambda compatibility test implementations for the Python suite.
"""

from __future__ import annotations
import base64
import json
import os
import time
import zipfile
import io
from lib.harness import TestContext
from lib.clients import make_clients

# Minimal Node.js Lambda handler — returns a simple JSON payload.
_HANDLER_JS = b"""
exports.handler = async (event) => ({
  statusCode: 200,
  body: JSON.stringify({ hello: "from lambda", event })
});
"""

# A Python Lambda handler for InvokeAsync tests.
_HANDLER_PY = b"""
def handler(event, context):
    return {"statusCode": 200, "body": "ok"}
"""

_ROLE_ARN = "arn:aws:iam::000000000000:role/lambda-role"
_LAYER_PY = b"""
# placeholder layer content
def layer_util():
    return "layer"
"""


def _make_zip(filename: str, content: bytes) -> bytes:
    buf = io.BytesIO()
    with zipfile.ZipFile(buf, "w", zipfile.ZIP_DEFLATED) as zf:
        zf.writestr(filename, content)
    return buf.getvalue()


def _lambda(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).lambda_


def _logs(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).logs


def _wait_active(ctx: TestContext, fn_name: str, max_attempts: int = 30) -> None:
    lmb = _lambda(ctx)
    for _ in range(max_attempts):
        resp = lmb.get_function(FunctionName=fn_name)
        state = resp.get("Configuration", {}).get("State")
        if state == "Active":
            return
        if state != "Pending":
            return
        time.sleep(0.2)
    raise AssertionError(f"Function {fn_name} did not become Active after {max_attempts} attempts")


def _delete_fn_and_logs(ctx: TestContext, fn_name: str) -> None:
    lmb = _lambda(ctx)
    logs = _logs(ctx)
    try:
        lmb.delete_function(FunctionName=fn_name)
    except Exception:
        pass
    try:
        logs.delete_log_group(logGroupName=f"/aws/lambda/{fn_name}")
    except Exception:
        pass


# ── lambda-crud ───────────────────────────────────────────────────────────────

def setup_lambda_crud(ctx: TestContext) -> None:
    fn_name = f"{ctx.run_id}-fn"
    _lambda(ctx).create_function(
        FunctionName=fn_name,
        Runtime="nodejs20.x",
        Role=_ROLE_ARN,
        Handler="index.handler",
        Code={"ZipFile": _make_zip("index.js", _HANDLER_JS)},
    )
    ctx["lambda_fn_name"] = fn_name


def teardown_lambda_crud(ctx: TestContext) -> None:
    fn_name = ctx.get("lambda_fn_name")
    if fn_name:
        _delete_fn_and_logs(ctx, fn_name)


def CreateFunction(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = f"{ctx.run_id}-fn-create"
    lmb.create_function(
        FunctionName=name,
        Runtime="nodejs20.x",
        Role=_ROLE_ARN,
        Handler="index.handler",
        Code={"ZipFile": _make_zip("index.js", _HANDLER_JS)},
    )
    try:
        resp = lmb.get_function(FunctionName=name)
        if resp["Configuration"]["FunctionName"] != name:
            raise AssertionError(f"CreateFunction: wrong name {resp['Configuration']['FunctionName']!r}")
    finally:
        lmb.delete_function(FunctionName=name)


def GetFunction(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_fn_name"]
    resp = lmb.get_function(FunctionName=name)
    cfg = resp.get("Configuration", {})
    if cfg.get("FunctionName") != name:
        raise AssertionError(f"GetFunction: wrong name {cfg.get('FunctionName')!r}")
    if not cfg.get("FunctionArn"):
        raise AssertionError("GetFunction: missing FunctionArn")


def ListFunctions(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_fn_name"]
    resp = lmb.list_functions()
    fns = [f["FunctionName"] for f in resp.get("Functions", [])]
    if name not in fns:
        raise AssertionError(f"ListFunctions: {name!r} not found in function list")


def UpdateFunctionCode(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_fn_name"]
    updated_js = b"""
exports.handler = async (event) => ({ statusCode: 200, body: 'updated' });
"""
    lmb.update_function_code(
        FunctionName=name,
        ZipFile=_make_zip("index.js", updated_js),
    )
    cfg = lmb.get_function_configuration(FunctionName=name)
    assert cfg.get("CodeSha256"), "UpdateFunctionCode: missing CodeSha256 after update"


def UpdateFunctionConfiguration(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_fn_name"]
    lmb.update_function_configuration(
        FunctionName=name,
        Description="updated by compat test",
        MemorySize=256,
    )
    resp = lmb.get_function_configuration(FunctionName=name)
    if resp.get("MemorySize") != 256:
        raise AssertionError(f"UpdateFunctionConfiguration: expected MemorySize=256, got {resp.get('MemorySize')}")


def DeleteFunction(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = f"{ctx.run_id}-fn-del"
    lmb.create_function(
        FunctionName=name,
        Runtime="nodejs20.x",
        Role=_ROLE_ARN,
        Handler="index.handler",
        Code={"ZipFile": _make_zip("index.js", _HANDLER_JS)},
    )
    lmb.delete_function(FunctionName=name)
    import botocore.exceptions
    try:
        lmb.get_function(FunctionName=name)
        raise AssertionError(f"DeleteFunction: function {name!r} still exists")
    except botocore.exceptions.ClientError as exc:
        if exc.response["Error"]["Code"] != "ResourceNotFoundException":
            raise


# ── lambda-invoke ─────────────────────────────────────────────────────────────

def setup_lambda_invoke(ctx: TestContext) -> None:
    fn_name = f"{ctx.run_id}-fn-invoke"
    _lambda(ctx).create_function(
        FunctionName=fn_name,
        Runtime="nodejs20.x",
        Role=_ROLE_ARN,
        Handler="index.handler",
        Code={"ZipFile": _make_zip("index.js", _HANDLER_JS)},
        Timeout=30,
    )
    ctx["lambda_invoke_fn"] = fn_name
    _wait_active(ctx, fn_name)


def teardown_lambda_invoke(ctx: TestContext) -> None:
    fn_name = ctx.get("lambda_invoke_fn")
    if fn_name:
        _delete_fn_and_logs(ctx, fn_name)


def InvokeDryRun(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_invoke_fn"]
    resp = lmb.invoke(FunctionName=name, InvocationType="DryRun")
    if resp["ResponseMetadata"]["HTTPStatusCode"] != 204:
        raise AssertionError(f"InvokeDryRun: expected 204, got {resp['ResponseMetadata']['HTTPStatusCode']}")


def InvokeSync(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_invoke_fn"]
    resp = lmb.invoke(
        FunctionName=name,
        InvocationType="RequestResponse",
        Payload=json.dumps({"test": True}).encode(),
    )
    if resp.get("FunctionError"):
        raise AssertionError(f"InvokeSync: function error {resp['FunctionError']!r}")
    payload = json.loads(resp["Payload"].read())
    if payload.get("statusCode") != 200:
        raise AssertionError(f"InvokeSync: unexpected payload {payload}")


def InvokeAsync(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_invoke_fn"]
    resp = lmb.invoke(
        FunctionName=name,
        InvocationType="Event",
        Payload=json.dumps({"async": True}).encode(),
    )
    if resp["ResponseMetadata"]["HTTPStatusCode"] != 202:
        raise AssertionError(f"InvokeAsync: expected 202, got {resp['ResponseMetadata']['HTTPStatusCode']}")


# ── lambda-invoke-stream ──────────────────────────────────────────────────────

def setup_lambda_invoke_stream(ctx: TestContext) -> None:
    fn_name = f"{ctx.run_id}-fn-stream"
    _lambda(ctx).create_function(
        FunctionName=fn_name,
        Runtime="nodejs20.x",
        Role=_ROLE_ARN,
        Handler="index.handler",
        Code={"ZipFile": _make_zip("index.js", _HANDLER_JS)},
        Timeout=30,
    )
    ctx["lambda_stream_fn"] = fn_name
    _wait_active(ctx, fn_name)


def teardown_lambda_invoke_stream(ctx: TestContext) -> None:
    fn_name = ctx.get("lambda_stream_fn")
    if fn_name:
        _delete_fn_and_logs(ctx, fn_name)


def InvokeWithResponseStream(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_stream_fn"]
    resp = lmb.invoke_with_response_stream(
        FunctionName=name,
        Payload=b'{}',
    )
    if resp["ResponseMetadata"]["HTTPStatusCode"] != 200:
        raise AssertionError(
            f"InvokeWithResponseStream: expected 200, got {resp['ResponseMetadata']['HTTPStatusCode']}"
        )
    # Consume the event stream.
    for event in resp.get("EventStream", []):
        pass


# ── lambda-aliases ────────────────────────────────────────────────────────────

def setup_lambda_aliases(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    fn_name = f"{ctx.run_id}-fn-alias"
    lmb.create_function(
        FunctionName=fn_name,
        Runtime="nodejs20.x",
        Role=_ROLE_ARN,
        Handler="index.handler",
        Code={"ZipFile": _make_zip("index.js", _HANDLER_JS)},
    )
    ctx["lambda_alias_fn"] = fn_name


def teardown_lambda_aliases(ctx: TestContext) -> None:
    fn_name = ctx.get("lambda_alias_fn")
    if fn_name:
        _delete_fn_and_logs(ctx, fn_name)


def PublishVersion(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_alias_fn"]
    resp = lmb.publish_version(FunctionName=name)
    if not resp.get("Version"):
        raise AssertionError("PublishVersion: missing Version in response")
    ctx["lambda_version"] = resp["Version"]


def ListVersionsByFunction(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_alias_fn"]
    resp = lmb.list_versions_by_function(FunctionName=name)
    versions = resp.get("Versions", [])
    if not versions:
        raise AssertionError("ListVersionsByFunction: no versions returned")


def CreateAlias(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_alias_fn"]
    version = ctx.get("lambda_version", "$LATEST")
    resp = lmb.create_alias(
        FunctionName=name,
        Name="live",
        FunctionVersion=version,
    )
    if not resp.get("AliasArn"):
        raise AssertionError("CreateAlias: missing AliasArn")
    ctx["lambda_alias_arn"] = resp["AliasArn"]


def GetAlias(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_alias_fn"]
    resp = lmb.get_alias(FunctionName=name, Name="live")
    if resp.get("Name") != "live":
        raise AssertionError(f"GetAlias: wrong name {resp.get('Name')!r}")


def ListAliases(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_alias_fn"]
    resp = lmb.list_aliases(FunctionName=name)
    aliases = [a["Name"] for a in resp.get("Aliases", [])]
    if "live" not in aliases:
        raise AssertionError(f"ListAliases: 'live' not found in {aliases}")


def UpdateAlias(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_alias_fn"]
    lmb.update_alias(FunctionName=name, Name="live", Description="updated by compat")
    resp = lmb.get_alias(FunctionName=name, Name="live")
    if resp.get("Description") != "updated by compat":
        raise AssertionError(f"UpdateAlias: description not updated; got {resp.get('Description')!r}")


def DeleteAlias(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_alias_fn"]
    lmb.delete_alias(FunctionName=name, Name="live")
    import botocore.exceptions
    try:
        lmb.get_alias(FunctionName=name, Name="live")
        raise AssertionError("DeleteAlias: alias still exists")
    except botocore.exceptions.ClientError as exc:
        if exc.response["Error"]["Code"] != "ResourceNotFoundException":
            raise


# ── lambda-layers ─────────────────────────────────────────────────────────────

def setup_lambda_layers(ctx: TestContext) -> None:
    ctx["lambda_layer_name"] = f"{ctx.run_id}-layer"


def teardown_lambda_layers(ctx: TestContext) -> None:
    pass  # layers deleted by tests


def PublishLayerVersion(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_layer_name"]
    resp = lmb.publish_layer_version(
        LayerName=name,
        Content={"ZipFile": _make_zip("lib.py", _LAYER_PY)},
        CompatibleRuntimes=["python3.12"],
    )
    if not resp.get("LayerArn"):
        raise AssertionError("PublishLayerVersion: missing LayerArn")
    ctx["lambda_layer_version"] = resp.get("Version", 1)


def ListLayers(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_layer_name"]
    resp = lmb.list_layers()
    layer_names = [l["LayerName"] for l in resp.get("Layers", [])]
    if name not in layer_names:
        raise AssertionError(f"ListLayers: {name!r} not found in {layer_names}")


def DeleteLayerVersion(ctx: TestContext) -> None:
    lmb = _lambda(ctx)
    name = ctx["lambda_layer_name"]
    version = ctx.get("lambda_layer_version", 1)
    lmb.delete_layer_version(LayerName=name, VersionNumber=version)
    resp = lmb.list_layer_versions(LayerName=name)
    versions = [v["Version"] for v in resp.get("LayerVersions", [])]
    assert version not in versions, f"DeleteLayerVersion: version {version} still present"


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "CreateFunction": CreateFunction,
    "GetFunction": GetFunction,
    "ListFunctions": ListFunctions,
    "UpdateFunctionCode": UpdateFunctionCode,
    "UpdateFunctionConfiguration": UpdateFunctionConfiguration,
    "DeleteFunction": DeleteFunction,
    "InvokeDryRun": InvokeDryRun,
    "InvokeSync": InvokeSync,
    "InvokeAsync": InvokeAsync,
    "InvokeWithResponseStream": InvokeWithResponseStream,
    "PublishVersion": PublishVersion,
    "ListVersionsByFunction": ListVersionsByFunction,
    "CreateAlias": CreateAlias,
    "GetAlias": GetAlias,
    "ListAliases": ListAliases,
    "UpdateAlias": UpdateAlias,
    "DeleteAlias": DeleteAlias,
    "PublishLayerVersion": PublishLayerVersion,
    "ListLayers": ListLayers,
    "DeleteLayerVersion": DeleteLayerVersion,
}

SETUP = {
    "lambda-crud": setup_lambda_crud,
    "lambda-invoke": setup_lambda_invoke,
    "lambda-invoke-stream": setup_lambda_invoke_stream,
    "lambda-aliases": setup_lambda_aliases,
    "lambda-layers": setup_lambda_layers,
}

TEARDOWN = {
    "lambda-crud": teardown_lambda_crud,
    "lambda-invoke": teardown_lambda_invoke,
    "lambda-invoke-stream": teardown_lambda_invoke_stream,
    "lambda-aliases": teardown_lambda_aliases,
    "lambda-layers": teardown_lambda_layers,
}

File: compat/suites/node-js-sdk/src/groups/lambda.ts
/**
 * groups/lambda.ts — Lambda compatibility test groups for the Node.js suite.
 *
 * Groups:
 *   lambda-crud    — function lifecycle (implemented)
 *   lambda-invoke  — synchronous invocation (requires Docker, skip if unavailable)
 *   lambda-invoke-stream — InvokeWithResponseStream (requires Docker, skip if unavailable)
 *   lambda-config  — environment variables, timeout, memory (implemented)
 *   lambda-aliases — versions and aliases (implemented)
 *   lambda-layers  — layer management (implemented)
 */

import {
  CreateFunctionCommand,
  DeleteFunctionCommand,
  GetFunctionCommand,
  ListFunctionsCommand,
  UpdateFunctionCodeCommand,
  UpdateFunctionConfigurationCommand,
  InvokeCommand,
  InvokeWithResponseStreamCommand,
  CreateAliasCommand,
  UpdateAliasCommand,
  GetAliasCommand,
  ListAliasesCommand,
  DeleteAliasCommand,
  PublishVersionCommand,
  ListVersionsByFunctionCommand,
  PublishLayerVersionCommand,
  ListLayersCommand,
  DeleteLayerVersionCommand,
  type FunctionConfiguration,
} from "@aws-sdk/client-lambda";
import { DeleteLogGroupCommand } from "@aws-sdk/client-cloudwatch-logs";
import { makeClients } from "../lib/clients.js";
import type { TestGroup } from "../lib/harness.js";
import * as assert from "node:assert/strict";

/**
 * Minimal Node.js Lambda handler bundled as a base64-encoded zip.
 * Handler: exports.handler = async () => ({ statusCode: 200, body: "ok" })
 * Generated programmatically with correct local header sizes and CRC-32.
 * (The previous zip was created by piping to `zip`, which left size=0 in
 * local headers and wrong offsets in the central directory, causing
 * Go's archive/zip to return ErrFormat on f.Open().)
 */
const MINIMAL_ZIP_BASE64 =
  "UEsDBBQAAAAAAAAAAAAKhksPNQAAADUAAAAIAAAAaW5kZXguanNleHBvcnRzLmhhbmRsZXI9YXN5bmMoKT0+KHtzdGF0dXNDb2RlOjIwMCxib2R5OiJvayJ9KVBLAQIUABQAAAAAAAAAAAAKhksPNQAAADUAAAAIAAAAAAAAAAAAAAAAAAAAAABpbmRleC5qc1BLBQYAAAAAAQABADYAAABbAAAAAAA=";

function makeZipBuffer(): Uint8Array {
  return Buffer.from(MINIMAL_ZIP_BASE64, "base64");
}

const ROLE_ARN = "arn:aws:iam::000000000000:role/lambda-exec";
const RUNTIME = "nodejs20.x";
const HANDLER = "index.handler";

import type { LambdaClient } from "@aws-sdk/client-lambda";
import type { CloudWatchLogsClient } from "@aws-sdk/client-cloudwatch-logs";

async function deleteFunctionAndLogGroup(
  lambda: LambdaClient,
  logs: CloudWatchLogsClient,
  name: string,
): Promise<void> {
  try { await lambda.send(new DeleteFunctionCommand({ FunctionName: name })); } catch {}
  try { await logs.send(new DeleteLogGroupCommand({ logGroupName: `/aws/lambda/${name}` })); } catch {}
}

async function waitFunctionActive(lambda: LambdaClient, name: string, maxAttempts = 30): Promise<void> {
  for (let i = 0; i < maxAttempts; i++) {
    const resp = await lambda.send(
      new GetFunctionCommand({ FunctionName: name }),
    );
    const state = resp.Configuration?.State;
    if (state === "Active") return;
    if (state !== "Pending") return;
    await new Promise((r) => setTimeout(r, 200));
  }
  throw new Error(`Function ${name} did not become Active after ${maxAttempts} attempts`);
}

export function makeLambdaGroups(suite: string): TestGroup[] {
  return [
    // ── lambda-crud ────────────────────────────────────────────────────────
    {
      suite,
      service: "lambda",
      name: "lambda-crud",
      tests: [
        {
          name: "CreateFunction",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const name = `${ctx.runId}-fn`;
            const resp = await lambda.send(
              new CreateFunctionCommand({
                FunctionName: name,
                Runtime: RUNTIME,
                Handler: HANDLER,
                Role: ROLE_ARN,
                Code: { ZipFile: makeZipBuffer() },
              }),
            );
            assert.ok(resp.FunctionArn, "CreateFunction: missing FunctionArn");
            assert.strictEqual(
              resp.FunctionName,
              name,
              `CreateFunction: expected name=${name}, got ${resp.FunctionName}`,
            );
            assert.ok(resp.CodeSha256, "CreateFunction: missing CodeSha256");
          },
        },
        {
          name: "GetFunction",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new GetFunctionCommand({ FunctionName: `${ctx.runId}-fn` }),
            );
            assert.ok(
              resp.Configuration?.FunctionArn,
              "GetFunction: missing FunctionArn",
            );
          },
        },
        {
          name: "ListFunctions",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const name = `${ctx.runId}-fn`;
            const resp = await lambda.send(new ListFunctionsCommand({}));
            if (
              !resp.Functions?.some(
                (f: FunctionConfiguration) => f.FunctionName === name,
              )
            ) {
              throw new Error(`ListFunctions: ${name} not found`);
            }
          },
        },
        {
          name: "UpdateFunctionCode",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new UpdateFunctionCodeCommand({
                FunctionName: `${ctx.runId}-fn`,
                ZipFile: makeZipBuffer(),
              }),
            );
            assert.ok(
              resp.FunctionArn,
              "UpdateFunctionCode: missing FunctionArn",
            );
            assert.ok(
              resp.CodeSha256,
              "UpdateFunctionCode: missing CodeSha256",
            );
          },
        },
        {
          name: "UpdateFunctionConfiguration",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new UpdateFunctionConfigurationCommand({
                FunctionName: `${ctx.runId}-fn`,
                Timeout: 30,
                MemorySize: 256,
                Environment: { Variables: { LOG_LEVEL: "debug" } },
              }),
            );
            assert.strictEqual(
              resp.Timeout,
              30,
              `UpdateFunctionConfiguration: expected Timeout=30, got ${resp.Timeout}`,
            );
          },
        },
        {
          name: "DeleteFunction",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const name = `${ctx.runId}-fn`;
            await lambda.send(
              new DeleteFunctionCommand({ FunctionName: name }),
            );
            const resp = await lambda.send(new ListFunctionsCommand({}));
            if (
              resp.Functions?.some(
                (f: FunctionConfiguration) => f.FunctionName === name,
              )
            ) {
              throw new Error(
                `DeleteFunction: ${name} still present after delete`,
              );
            }
          },
        },
      ],
      teardown: async (ctx) => {
        const { lambda, logs } = makeClients(ctx);
        await deleteFunctionAndLogGroup(lambda, logs, `${ctx.runId}-fn`);
      },
    },

    // ── lambda-invoke ──────────────────────────────────────────────────────
    {
      suite,
      service: "lambda",
      name: "lambda-invoke",
      tests: [
        {
          name: "InvokeDryRun",
          op: "Invoke",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new InvokeCommand({
                FunctionName: `${ctx.runId}-fn-invoke`,
                InvocationType: "DryRun",
              }),
            );
            assert.strictEqual(
              resp.StatusCode,
              204,
              `InvokeDryRun: expected StatusCode=204, got ${resp.StatusCode}`,
            );
          },
        },
        {
          name: "InvokeSync",
          // Real Docker invocation — skip if OVERCAST_COMPAT_SKIP_DOCKER is set
          skip:
            process.env.OVERCAST_COMPAT_SKIP_DOCKER === "1"
              ? "Docker not available (set OVERCAST_COMPAT_SKIP_DOCKER=0 to enable)"
              : false,
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new InvokeCommand({
                FunctionName: `${ctx.runId}-fn-invoke`,
                InvocationType: "RequestResponse",
                Payload: new TextEncoder().encode(
                  JSON.stringify({ test: true }),
                ),
              }),
            );
            assert.strictEqual(
              resp.StatusCode,
              200,
              `InvokeSync: expected StatusCode=200, got ${resp.StatusCode}`,
            );
            if (resp.FunctionError) {
              const body = resp.Payload
                ? new TextDecoder().decode(resp.Payload)
                : "";
              throw new Error(
                `InvokeSync: function error: ${resp.FunctionError} — ${body}`,
              );
            }
          },
        },
        {
          name: "InvokeAsync",
          skip:
            process.env.OVERCAST_COMPAT_SKIP_DOCKER === "1"
              ? "Docker not available (set OVERCAST_COMPAT_SKIP_DOCKER=0 to enable)"
              : false,
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new InvokeCommand({
                FunctionName: `${ctx.runId}-fn-invoke`,
                InvocationType: "Event",
                Payload: new TextEncoder().encode("{}"),
              }),
            );
            assert.strictEqual(
              resp.StatusCode,
              202,
              `InvokeAsync: expected StatusCode=202, got ${resp.StatusCode}`,
            );
          },
        },
      ],
      setup: async (ctx) => {
        const { lambda } = makeClients(ctx);
        const name = `${ctx.runId}-fn-invoke`;
        await lambda.send(
          new CreateFunctionCommand({
            FunctionName: name,
            Runtime: RUNTIME,
            Handler: HANDLER,
            Role: ROLE_ARN,
            Code: { ZipFile: makeZipBuffer() },
            Timeout: 30,
          }),
        );
        await waitFunctionActive(lambda, name);
      },
      teardown: async (ctx) => {
        const { lambda, logs } = makeClients(ctx);
        await deleteFunctionAndLogGroup(lambda, logs, `${ctx.runId}-fn-invoke`);
      },
    },

    // ── lambda-invoke-stream ───────────────────────────────────────────────
    {
      suite,
      service: "lambda",
      name: "lambda-invoke-stream",
      tests: [
        {
          name: "InvokeWithResponseStream",
          skip:
            process.env.OVERCAST_COMPAT_SKIP_DOCKER === "1"
              ? "Docker not available (set OVERCAST_COMPAT_SKIP_DOCKER=0 to enable)"
              : false,
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new InvokeWithResponseStreamCommand({
                FunctionName: `${ctx.runId}-fn-stream`,
                InvocationType: "RequestResponse",
                Payload: new TextEncoder().encode(
                  JSON.stringify({ test: true }),
                ),
              }),
            );
            assert.strictEqual(
              resp.StatusCode,
              200,
              `InvokeWithResponseStream: expected StatusCode=200, got ${resp.StatusCode}`,
            );
            assert.strictEqual(
              resp.ExecutedVersion,
              "$LATEST",
              `InvokeWithResponseStream: expected ExecutedVersion=$LATEST`,
            );

            // Consume event stream and collect chunks.
            const chunks: Uint8Array[] = [];
            let completed = false;
            if (resp.EventStream) {
              for await (const event of resp.EventStream) {
                if (event.PayloadChunk?.Payload) {
                  chunks.push(event.PayloadChunk.Payload);
                }
                if (event.InvokeComplete !== undefined) {
                  completed = true;
                }
              }
            }
            assert.ok(
              completed,
              "InvokeWithResponseStream: expected InvokeComplete event",
            );
            assert.ok(
              chunks.length > 0,
              "InvokeWithResponseStream: expected at least one PayloadChunk",
            );
          },
        },
      ],
      setup: async (ctx) => {
        const { lambda } = makeClients(ctx);
        const name = `${ctx.runId}-fn-stream`;
        await lambda.send(
          new CreateFunctionCommand({
            FunctionName: name,
            Runtime: RUNTIME,
            Handler: HANDLER,
            Role: ROLE_ARN,
            Code: { ZipFile: makeZipBuffer() },
            Timeout: 30,
          }),
        );
        await waitFunctionActive(lambda, name);
      },
      teardown: async (ctx) => {
        const { lambda, logs } = makeClients(ctx);
        await deleteFunctionAndLogGroup(lambda, logs, `${ctx.runId}-fn-stream`);
      },
    },

    // ── lambda-aliases ─────────────────────────────────────────────────────
    {
      suite,
      service: "lambda",
      name: "lambda-aliases",
      tests: [
        {
          name: "PublishVersion",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new PublishVersionCommand({
                FunctionName: `${ctx.runId}-fn-alias`,
              }),
            );
            assert.ok(resp.Version, "PublishVersion: missing Version");
            ctx["_version"] = resp.Version;
          },
        },
        {
          name: "ListVersionsByFunction",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new ListVersionsByFunctionCommand({
                FunctionName: `${ctx.runId}-fn-alias`,
              }),
            );
            assert.notStrictEqual(
              resp.Versions?.length ?? 0,
              0,
              "ListVersionsByFunction: expected at least one version",
            );
          },
        },
        {
          name: "CreateAlias",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const version = (ctx["_version"] as string) ?? "1";
            const resp = await lambda.send(
              new CreateAliasCommand({
                FunctionName: `${ctx.runId}-fn-alias`,
                Name: "live",
                FunctionVersion: version,
              }),
            );
            assert.ok(resp.AliasArn, "CreateAlias: missing AliasArn");
          },
        },
        {
          name: "GetAlias",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new GetAliasCommand({
                FunctionName: `${ctx.runId}-fn-alias`,
                Name: "live",
              }),
            );
            assert.strictEqual(
              resp.Name,
              "live",
              `GetAlias: expected Name=live, got ${resp.Name}`,
            );
          },
        },
        {
          name: "ListAliases",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new ListAliasesCommand({ FunctionName: `${ctx.runId}-fn-alias` }),
            );
            assert.ok(
              resp.Aliases?.some((a) => a.Name === "live"),
              "ListAliases: expected alias 'live'",
            );
          },
        },
        {
          name: "UpdateAlias",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(
              new UpdateAliasCommand({
                FunctionName: `${ctx.runId}-fn-alias`,
                Name: "live",
                Description: "production alias",
              }),
            );
            assert.strictEqual(
              resp.Description,
              "production alias",
              `UpdateAlias: expected Description="production alias"`,
            );
          },
        },
        {
          name: "DeleteAlias",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            await lambda.send(
              new DeleteAliasCommand({
                FunctionName: `${ctx.runId}-fn-alias`,
                Name: "live",
              }),
            );
            const { Aliases = [] } = await lambda.send(
              new ListAliasesCommand({ FunctionName: `${ctx.runId}-fn-alias` }),
            );
            assert.notStrictEqual(
              Aliases.some((a) => a.Name, "live"),
              "DeleteAlias: alias 'live' still present after delete",
            );
          },
        },
      ],
      setup: async (ctx) => {
        const { lambda } = makeClients(ctx);
        await lambda.send(
          new CreateFunctionCommand({
            FunctionName: `${ctx.runId}-fn-alias`,
            Runtime: RUNTIME,
            Handler: HANDLER,
            Role: ROLE_ARN,
            Code: { ZipFile: makeZipBuffer() },
          }),
        );
      },
      teardown: async (ctx) => {
        const { lambda, logs } = makeClients(ctx);
        await deleteFunctionAndLogGroup(lambda, logs, `${ctx.runId}-fn-alias`);
      },
    },

    // ── lambda-layers ──────────────────────────────────────────────────────
    {
      suite,
      service: "lambda",
      name: "lambda-layers",
      tests: [
        {
          name: "PublishLayerVersion",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const layerName = `${ctx.runId}-layer`;
            const resp = await lambda.send(
              new PublishLayerVersionCommand({
                LayerName: layerName,
                Description: "test layer",
                Content: { ZipFile: makeZipBuffer() },
                CompatibleRuntimes: [RUNTIME],
              }),
            );
            assert.ok(
              resp.LayerVersionArn,
              "PublishLayerVersion: missing LayerVersionArn",
            );
            ctx["_layerArn"] = resp.LayerVersionArn;
            ctx["_layerVersion"] = resp.Version;
          },
        },
        {
          name: "ListLayers",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const resp = await lambda.send(new ListLayersCommand({}));
            const layerName = `${ctx.runId}-layer`;
            assert.ok(
              resp.Layers?.some((l) => l.LayerName === layerName),
              `ListLayers: ${layerName} not found`,
            );
          },
        },
        {
          name: "DeleteLayerVersion",
          fn: async (ctx) => {
            const { lambda } = makeClients(ctx);
            const layerName = `${ctx.runId}-layer`;
            const version = (ctx["_layerVersion"] as number) ?? 1;
            await lambda.send(
              new DeleteLayerVersionCommand({
                LayerName: layerName,
                VersionNumber: version,
              }),
            );
          },
        },
      ],
      teardown: async (ctx) => {
        const { lambda } = makeClients(ctx);
        const layerName = `${ctx.runId}-layer`;
        const version = (ctx["_layerVersion"] as number) ?? 1;
        try {
          await lambda.send(
            new DeleteLayerVersionCommand({
              LayerName: layerName,
              VersionNumber: version,
            }),
          );
        } catch {}
      },
    },
  ];
}
